### ISSUE TITLE
Unnecessary division by dt?

### ISSUE THREAD

**Issue by fawidmer** [jhelgert/pyCIAP#2]
Hello

I've been going through your code and I believe to have found a minor mistake:
* In the `DSUR` function, you expect the `min_up_time` and `int min_down_time` to be given as an integer (number of time steps).
* However, in the function `down_time_forbidden_configs`, you then divide the `down_time` by `dt`, as if `down_time` were given in number of seconds
* I think simply setting `e_c = down_time` would be correct?

**Comment by jhelgert** [comment]
Hey, thanks for your issue. Yes, you're right, the division by `dt` is not needed at all since the minimum up/down dwell times are given as integer time steps. 

That being said, my (old) implementation is far from perfect and could be simplified in a lot of places as it was a straightforward implementation of the algorithm presented in the paper. I guess it's worth mentioning that I refactored the code a while ago. However, I've never found the time to create a PR since then, unfortunately.

> I think simply setting e_c = down_time would be correct?

Yeah, I think so. You may also compare it to [this](https://gist.github.com/jhelgert/25c93fa561b59c2af48acd9c622c8953) refactored and simplified python only implementation of the algorithm, which is way easier to read, IMO. Hope this helps!

**Comment by fawidmer** [comment]
Dear Jonathan
Thank you so much for your immediate response!! And thank you for the refactored code.
One thing that I still do not quite understand:
Testing the code (the old or the new one) with this very trivial example:
```python
b_rel = np.array([[1, 1, 1, 1, 1, 1, 1, 1],
    [0, 0, 0, 0, 0, 0, 0, 0]])
```

Both implementations return the following:
```
[[1 1 1 0 0 0 1 1]
 [0 0 0 1 1 1 0 0]]
```

This is surprising to me, since the input is already integer and fulfills the MU and MD constraints? There are two options to resolve my confusion, I hope you can clarify :)

1. My understanding of the minimum down time constraint is wrong. I understand it as follows: If the control switches from w_0 to any other control, then w_0 cannot be activated for at least the time specified by the MDT constraint. If this is correct, the variable `b_rel` above already satisfies this, since it never deactivates any control.
1. There is a mistake in the implementation. I suspect it would be in `down_time_forbidden_configs` (`update_configs` in the new implementation).

Thank you so much for your help, I really appreciate it!!

Fabio

**Comment by jhelgert** [comment]
Your understanding of the minimum down time constraint is correct. However, your trivial example contradicts the algorithm's assumption that the passed control `b_rel` (`alpha` resp.) is a relaxed control, i.e. a control that isn't binary feasible. Your trivial example is already binary feasible, so I think the implementation behaves as expected in this case. Of course, it is still possible that the code contains bugs. If you find any, please let me know :).

PS: If you don't want to stick to the DSUR algorithm, you might also take a look at [pycombina](https://github.com/adbuerger/pycombina). This package can solve the same kind of problems but it uses a tailored B&B algorithm and supports more constraints.

**Comment by fawidmer** [comment]
Good point!

However, I changed my minimum example to the following

```python
vector = np.array([[1 - 1e-5],
    [1e-5]])

b_rel = np.tile(vector, (1, 20))
```

-> This solution is "almost" binary feasible.

However, the algorithm still returns something similar as before:
```
[[1 1 1 0 0 0 1 1 1 0 0 0 1 1 1 0 0 0 1 1]
 [0 0 0 1 1 1 0 0 0 1 1 1 0 0 0 1 1 1 0 0]]
```

So I guess there really is a bug somewhere?

I could try to find and fix the bug, but I should have a way to test the corrected solution ... Do you have a few sample input-output data pairs that you know should be correct that I could compare any implementation to?

**Comment by jhelgert** [comment]
> So I guess there really is a bug somewhere?

Yes, indeed, good catch! Thanks a lot for your example. I had a quick look at it and it seems like the bug is inside the `update_configs` function, i.e. I used wrong indices there. Would you mind giving [this one](https://gist.github.com/jhelgert/25c93fa561b59c2af48acd9c622c8953) another try? I updated it and it seems to yield reasonable results now. Disclaimer: I only tested it for 2-3 simple examples, so it's not tested properly. 

> Do you have a few sample input-output data pairs that you know should be correct that I could compare any implementation to?

I'm afraid I don't :/

**Comment by fawidmer** [comment]
Thanks a lot!
Indeed, the results for my minimum example certainly make more sense now.
However, I think it's still not quite correct now, since you are now looking at indices "in the future" when using `start : start + num_dwell_steps`. However, within the algorithm, those values will always be zero, so I think you will never have any disallowed controls.

However, at least in the case with only two modes, I could not quickly find an example where this might become an issue. For instance, I was first expecting this example to neglect the minimum down time constraint:
```python
vector = np.array([[1 - 2e-1],
                       [2e-1]])

b_rel = np.tile(vector, (1, 30))

b = sum_up_rounding(b_rel, min_up_time_steps=3, min_down_time_steps=6)
```

But it works just fine, producing this result:
```
[[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1]
 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0]]
```

**Comment by jhelgert** [comment]
> Thanks a lot! Indeed, the results for my minimum example certainly make more sense now. However, I think it's still not quite correct now, since you are now looking at indices "in the future" when using `start : start + num_dwell_steps`. However, within the algorithm, those values will always be zero, so I think you will never have any disallowed controls.

Yeah, good point. I guess we'd need to compare it again with the proposed algorithm in the paper[^1] and find some examples where the implementation doesn't yield the expected results compared to the original algorithm. However, I'm quite busy for the next weeks, so I can't promise anything, unfortunately :/. However, I'm happy to review any PR in case you want to have a look at it. In case you need a stable alternative, I'd propose to use pycombina instead.

[^1]: It's algorithm 2 in https://link.springer.com/content/pdf/10.1007/s10107-020-01533-x.pdf?pdf=button

**Comment by fawidmer** [comment]
Dear Jonathan
Thanks a lot for your help!
For now, I can work with the version that you've sent me, since I only have two discrete states.
